python/ray/_private/log_rotation_with_pipe.py

```diff
@@ -0,0 +1,142 @@
+"""Implement log rotation with pipe."""
+
+from typing import IO, AnyStr
+import sys
+import os
+import logging
+import threading
+import uuid
+import atexit
+from logging.handlers import RotatingFileHandler
+
+# Use a special token to indicate write finish.
+_EOF_TOKEN = str(uuid.uuid4()) + "\n"
+
+
+class PipeStreamWriteHandle(IO[AnyStr]):
+    def __init__(self, write_fd: int, log_name: str):
+        self.name = log_name
+        self.write_fd = write_fd
+        self.stream = os.fdopen(write_fd, "w")
+
+    def name(self):
+        return self.name
+
+    def fileno(self):
+        return self.write_fd
+
+    def write(self, s: str) -> int:
+        return self.stream.write(s)
+
+    def flush(self):
+        self.stream.flush()
+
+    def close(self):
+        self.stream.close()
+
+    def isatty(self) -> bool:
+        return False
+
+    def readable(self) -> bool:
+        return False
+
+    def writable(self) -> bool:
+        return True
+
+    def seekable(self) -> bool:
+        return False
+
+
+def open_pipe_with_file_handlers(log_name: str, handlers) -> IO[AnyStr]:
+    """Stream content into pipe, which will be listened and dumped to provided
+    handlers."""
+    read_fd, write_fd = os.pipe()
+
+    log_content = []
+    stopped = threading.Event()
+    cond = threading.Condition()
+
+    # Have to use different logger names for different sinks.
+    logger = logging.getLogger(f"logger-{log_name}")
+    logger.setLevel(logging.INFO)
+    for handler in handlers:
+        logger.addHandler(handler)
+
+    # Setup read thread, which continuous read content out of pipe and append to buffer.
+    def read_log_content_from_pipe():
+        with os.fdopen(read_fd, "r") as pipe_reader:
+            while True:
+                line = pipe_reader.readline()
+
+                with cond:
+                    # An empty line is read over, which indicates write side has
+                    # closed.
+                    if line == _EOF_TOKEN:
```

Comment:
why do we need `_EOF_TOKEN`?  If you close the pipe, reader will know? `readline()` should return empty string to indicate EOF.